src/v/model/tests/ktp_test.cc

```diff
@@ -0,0 +1,274 @@
+// Copyright 2020 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "model/fundamental.h"
+#include "model/ktp.h"
+
+#include <seastar/core/sstring.hh>
+
+#include <boost/test/tools/old/interface.hpp>
+#include <boost/test/unit_test.hpp>
+#include <boost/test/unit_test_suite.hpp>
+
+using model::ktp;
+using model::ktp_with_hash;
+using model::ntp;
+
+static ktp makek(const ss::sstring& t, const int p) {
+    return ktp{model::topic(t), model::partition_id(p)};
+}
+
+static ktp_with_hash makekh(const ss::sstring& t, const int p) {
+    return model::ktp_with_hash{model::topic(t), model::partition_id(p)};
+}
+
+static ntp maken(const ss::sstring& t, const int p) {
+    return ntp{model::kafka_namespace, model::topic(t), model::partition_id(p)};
+}
+
+template<typename T>
+static size_t hash_of(const T& t) {
+    return std::hash<T>{}(t);
+}
+
+static size_t hashk(const ss::sstring& t, const int p) {
+    return hash_of(makek(t, p));
+}
+
+static size_t hashkh(const ss::sstring& t, const int p) {
+    return hash_of(makekh(t, p));
+}
+
+static size_t hashn(const ss::sstring& t, const int p) {
+    return hash_of(maken(t, p));
+}
+
+template<typename L, typename R>
+void equals_helper_inner(L makel, R maker) {
+    BOOST_CHECK_EQUAL(makel("", 0), maker("", 0));
+    BOOST_CHECK_EQUAL(makel("topic", 0), maker("topic", 0));
+    BOOST_CHECK_EQUAL(makel("topic", 100), maker("topic", 100));
+
+    BOOST_CHECK_NE(makel("", 0), maker("", 1));
+    BOOST_CHECK_NE(makel("", 1), maker("", 0));
+    BOOST_CHECK_NE(makel("a", 0), maker("", 0));
+    BOOST_CHECK_NE(makel("", 0), maker("a", 0));
+
+    BOOST_CHECK(makel("", 0) != maker("", 1)); // check != operator
+}
+
+template<typename L, typename R>
+void equals_helper(L make1, R make2) {
+    equals_helper_inner(make1, make2);
+    equals_helper_inner(make2, make1);
+}
+
+BOOST_AUTO_TEST_CASE(test_ktp_equals_default_objects) {
+    BOOST_CHECK_EQUAL(ktp(), ktp());
+    BOOST_CHECK_EQUAL(ktp(), ktp_with_hash());
+    BOOST_CHECK_EQUAL(ktp_with_hash(), ktp_with_hash());
+}
+
+BOOST_AUTO_TEST_CASE(test_ktp_equals_internal) { equals_helper(makek, makek); }
+
+BOOST_AUTO_TEST_CASE(test_ktp_equals_cross_with_hash) {
+    equals_helper(makek, makekh);
+}
+
+BOOST_AUTO_TEST_CASE(test_equals_cross) { equals_helper(makek, maken); }
```

Comment:
C++20 rewrites == and other relational operators so that for x == y, it will also check effectively for something that would match y == x, so you don't need to write all these mostly redundant overloads.

Reply:
@dlex resolving this and assuming you are OK with this explanation.